usb: device_next: add initial HID device support

jfischer-no opened the pull request:
Initial HID device support.
~Depends on a few patches from different PRs.~

jfischer-no committed:
usb: device_next: add initial HID device support

Add initial HID device support. Unlike the existing HID implementation,
the new implementation uses a devicetree to instantiate a HID device.
To the user, the HID device appears as a normal Zephyr RTOS device.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: allow Get Descriptor with recipient interface

If the recipient is not the device then it is probably a
class specific  request where wIndex is the interface
number or endpoing and not the language ID. e.g. HID
Class Get Descriptor request.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: hid-mouse: add support for the new HID implementation

Add support for the new HID implementation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
include: usb_hid: remove conditional definition from struct hid_ops

For upcoming HID support in the new device stack and to provide API
backward compatibility, there must be no conditional definitions in the
header.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on subsys/usb/device_next/usbd_ch9.c:
```diff
@@ -458,6 +461,16 @@ static int sreq_get_descriptor(struct usbd_contex *const uds_ctx,
+		/*
+		 * If the recipient is not the device then it is probably a
+		 * class specific  request where wIndex is the interface
+		 * number or endpoing and not the language ID. e.g. HID
```
`endpoing` -> `endpoint`

tmon-nordic commented on dts/bindings/usb/zephyr,hid-device.yaml:
```diff
@@ -0,0 +1,50 @@
+    type: int
+    required: true
+    description: |
+      Input type reports polling rate in number of 125µs periods.
+      Depending on the actual speed, the calculated polling rate could be clamped
+      to 1ms or 255ms.
```
The 125µs period is the case only on high-speed. On full-speed the meaning is different.

XenuIsWatching replied:
I don't think it's exactly that either... On High Speed it's the formula of 2^(bInterval-1) = x (125ms). Where it would follow this table. Only Values of 1 to 16 are valid.
| bInterval | uS      |
|-----------|---------|
| 1         | 125     |
| 2         | 250     |
| 3         | 500     |
| 4         | 1000    |
| 5         | 2000    |
| 6         | 4000    |
| 7         | 8000    |
| 8         | 16000   |
| 9         | 32000   |
| 10        | 64000   |
| 11        | 128000  |
| 12        | 256000  |
| 13        | 512000  |
| 14        | 1024000 |
| 15        | 2048000 |
| 16        | 4096000 |

For Full/Low Speed it's, just 1ms units.

jfischer-no replied:
Thanks for the point. I want to use 125µs granularity for this property. But the actual code does not handle it properly.

tmon-nordic commented on dts/bindings/usb/zephyr,hid-device.yaml:
```diff
@@ -0,0 +1,50 @@
+      Depending on the actual speed, the calculated polling rate could be clamped
+      to 1ms or 255ms.
+
+  in-reports-mps:
```
I think we should have separate max packet size entry for full-speed and high-speed, because the two have different allowed ranges. While the USB stack does not really support the concept of different configurations for different speed we really should have it because otherwise we'll run into unsolvable issues later on.

tmon-nordic reviewed (changes requested):
Must not be merged before #65165 and #65457

jfischer-no committed:
samples: usb: add HID keyboard sample

Add HID keyboard sample for the new experimental USB device support.
This is a limited and not fully compliant HID keyboard implementation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on samples/subsys/usb/hid-keyboard/Kconfig:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2023 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+# Source common USB sample options used to initialize new experimental USB
+# device stack. You do not need to source this Kconfig file in your custom
+# application.
```
Align this with the text in main that says this must not be used in applications.

benedekkupper commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,721 @@
+	}
+
+	switch (type) {
+	case HID_REPORT_TYPE_INPUT:
```
Input report type cannot be set, this should be an error case.

jfischer-no replied:
Maybe I missed another ECN, but from the USB HID spec, both Get Report and Set Report are valid for all three report types. 7.2.1 "This request is useful at initialization time for absolute items and for determining the state of feature items. This request is not intended to be used for polling the device state on a regular basis." and 7.2.2 "The Set_Report request allows the host to send a report to the device, possibly setting the state of input, output, or feature controls."

benedekkupper replied:
The same document talks about Input, Output and Feature reports in a different light in the Remarks at 6.2.2.5:
> Input items define input reports accessible via the Control pipe with a Get_Report (Input) request.

> Output items make Output reports accessible via the Control pipe with a Set_Report (Output) command.

> Output type reports can optionally be sent via an Interrupt Out pipe.

> Like Output items, Feature items make up Feature Reports accessible via the Control pipe with the Get_Report (Feature) and Set_Report (Feature) requests.

At the same time, this document is mixing these terms (Input, Output, Feature) between a category on the protocol level versus a category on the device - human interface level, the latter being a lot more permissive. All in all I think you are reading it right and sticking to the specification it should be allowed - even if it's very unlikely that any host OS ever does this. Maybe it would be worth it to try with a rawhid device and check if hidapi allows doing this.

tmon-nordic replied:
I think this is incorrect interpretation. The HID specification does explicitly mention `Set_Report(Output)` and `Set_Report(Feature)` while it never mentions an input version.

`6.2.2.5 Input, Output, and Feature Items` states:
> * An Input item describes information about the data provided by one or more physical controls. An application can use this information to interpret the data provided by the device. All data fields defined in a single item share an identical data format.
> * The Output item is used to define an output data field in a report. This item is similar to an Input item except it describes data sent to the device—for example, LED states.
> * Feature items describe device configuration information that can be sent to the device.

and later in the Remarks:
> * Input items define input reports accessible via the Control pipe with a Get_Report (Input) request.

> * Output items make Output reports accessible via the Control pipe with a Set_Report (Output) command.

> * Feature items may be attributes of other items. For example, an Origin Reset Feature may apply to one or more position Input items. Like Output items, Feature items make up Feature Reports accessible via the Control pipe with the Get_Report (Feature) and Set_Report (Feature) requests.

tmon-nordic replied:
Additionally HID Usage Tables Terms and Abbreviations define Application as:
> A software program that consumes the data generated by the HID device Input reports, or that controls the HID device through Feature or Output reports. Applications can be games or other programs used by end users or system software components.

Therefore it explicitly mentions that Input is from device to host, and only Feature or Output reports can control the device (from host to device). This matches the bullet points in HID Specification 6.2.2.5 Input, Output, and Feature Items.

jfischer-no replied:
This is not about how reports are intended to be used, and the current implementation does not force anyone to use all the parameters in these callbacks. The spec explicitly mentions all three report types in the description of Get Report and Set Report. Set Report is even more specific, see 7.2.2,
"
```
A device might choose to ignore input Set_Report requests as meaningless.
Alternatively these reports could be used to reset the origin of a control (that
is, current position should report zero). The effect of sent reports will also
depend on whether the recipient controls are absolute or relative.
```
"

benedekkupper commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,721 @@
+		errno = ops->get_report(dev, type, id, net_buf_tailroom(buf), buf->data);
+		break;
+	case HID_REPORT_TYPE_OUTPUT:
+		LOG_DBG("Get Report, Output Report ID %u", id);
```
Output report type cannot be get, this should be an error case.

tmon-nordic replied:
Just like with input report being set, HID Specification leaves it up to developer whether it is possible to get output report or not.

benedekkupper commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,721 @@
+	const struct hid_device_ops *const ops = ddata->ops;
+	struct usbd_hid_descriptor *const desc = ddata->desc;
+
+	atomic_set_bit(&ddata->state, HID_DEV_CLASS_ENABLED);
```
I'm not sure where to put this, and maybe I just didn't find where this happens, but `ddata->protocol` should be reset to REPORT when the HID interface is configured.

MarekPieta replied:
Maybe we could also call application's callback informing about the protocol change here?

jfischer-no replied:
below, ready callback is called. Also host should set the protocol whenever initializing a device.

MarekPieta replied:
So that application is expected to switch to report protocol on ready callback? Sounds ok too.

tmon-nordic commented on include/zephyr/usb/class/usbd_hid.h:
```diff
@@ -0,0 +1,209 @@
+
+	/**
+	 * Callback to set a feature, input, or output request, specified by
+	 * type an id, from the not to the HID device. This callback is called
```
While the "type an id" seems to be simply missing "d", i.e. should be "type and id", I cannot quite grasp what "from the not to the HID device" is supposed to mean.

tmon-nordic reviewed (changes requested):
The latest update didn't address previous review comments.

benedekkupper commented on include/zephyr/usb/class/usbd_hid.h:
```diff
@@ -0,0 +1,207 @@
+	 */
+	void (*input_report_done)(const struct device *dev);
+
+	/**
+	 * New output report callback. Callback will only be called for reports
+	 * received through the optional interrupt OUT pipe. If there is no
+	 * interrupt OUT pipe, output reports will be received using set_report().
+	 * If a report ID is used in the report descriptor, the host places the ID
+	 * in the buffer first, followed by the report data.
+	 */
+	void (*output_report)(const struct device *dev, const uint16_t len,
+			      const uint8_t *const buf);
```
Why not feed the data received through the OUT interrupt pipe also through the set_report callback instead? The upper level HID application should be agnostic to the detail of which pipe the output report arrived on, especially if the intention is to make this API transport independent.

jfischer-no replied:
I had changed it to use set_report, not sure if the commit was published in that PR, but changed it back because I found it confusing. In the Get/Set Report, there is an explicit report id parameter. Report id is optional, but then the id parameter is reserved 0 and everything is clear. For the input report, the IN endpoint should be used, although the input report can also be retrieved using Get Report, the presentation is slightly different. So different representations -> different interfaces. For output reports, I find it more intuitive to make them similar to input report interface and not abuse Set Report with ID 0. So it can be useful to know which interface is being used, passing it through the same callback as Set Report will remove this information. The application, like keyboard sample, can just call the same input report code with ID 0 if there are no IDs in the report descriptor.

benedekkupper replied:
Well if there is a report ID parameter, that report ID will also occupy the first byte of the transferred report, so passing the report ID as an extra argument in the set_report is redundant. The same high level application might be ported to different targets with or without a dedicated OUT endpoint, which might create some inconvenience.

How about keeping the separate callback, but if it's not implemented by the application, the driver falls back to calling the set_report?

jfischer-no replied:
> How about keeping the separate callback, but if it's not implemented by the application, the driver falls back to calling the set_report?

I am very much against it. Having or not having an optional OUT endpoint is not a bagatelle. The HID device implementor needs to be aware of the differences. If there is no need for low latency output reports then OUT endpoint should not be used. Also, something like "HID RAW" for regular data transfer should not be used and is discouraged. Btw, ID parameter in set_report is not redundant, interface is different, type and id in set_report are from setup packet, report "payload" from the data stage.

ddavidebor replied:
> I am very much against it. Having or not having an optional OUT endpoint is not a bagatelle. The HID device implementor needs to be aware of the differences. If there is no need for low latency output reports then OUT endpoint should not be used.

I agree with @jfischer-no , it should be explicit, not a fallback. 

I can see that @benedekkupper would like a more polished, higher-lever HID experience, probably thinking about HID over GATT + HID over USB. That's great as it's own thing, but for the USB, it needs to stick to the way USB HID works, or it will cause a lot of headaches.

benedekkupper replied:
> Btw, ID parameter in set_report is not redundant, interface is different, type and id in set_report are from setup packet, report "payload" from the data stage.

They are redundant in the sense that the spec requires them to be the same, even if the same values appears twice on the bus.

> Also, something like "HID RAW" for regular data transfer should not be used and is discouraged.

How do you define "regular data transfer"? HID has seen a lot of use as a general purpose middleware transport, just on top of my head there's FIDO, CMSIS DAP and NXP's KBOOT flashing protocols.

In the end I'm ok with the current situation, I understand the justifications now.

jfischer-no replied:
> How do you define "regular data transfer"? HID has seen a lot of use as a general purpose middleware transport, just on top of my head there's FIDO, CMSIS DAP and NXP's KBOOT flashing protocols.

By regular, I mean something that uses reports as a data payload and does not require periodic transfers. CMSIS-DAP moved to bulk transfers, e.g. discussion in https://github.com/zephyrproject-rtos/zephyr/pull/53798

benedekkupper commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+
+#define HID_SUBORDINATE_DESC_NUM		1
+
+struct subordiante_info {
```
typo subordiante?

benedekkupper commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+		duration = ops->get_idle(dev, id) / 4UL;
+	}
+
+	LOG_DBG("Get Idle, Report ID %u Durationi %u", id, duration);
```
typo Durationi

benedekkupper commented on include/zephyr/usb/class/usbd_hid.h:
```diff
@@ -0,0 +1,207 @@
+	/**
+	 * Notification to limit intput report frequency.
+	 * The application should mute an input report submission until a new
+	 * event occurs or until the time specified by the duration value has
```
Two things here:
1. It would be nice to mention that the duration is expressed in milliseconds (as opposed to the raw USB value which has 4ms resolution).
2. It would be nice to group the USB-only part of the HID API (set_idle, get_idle, sof) separately, even a prefix would be nice, just as an indication of transport dependent support for these.

jfischer-no replied:
> Two things here:
> 
>     1. It would be nice to mention that the duration is expressed in milliseconds (as opposed to the raw USB value which has 4ms resolution).

I removed */4U, it is 4 millisecond resolution now.

>     2. It would be nice to group the USB-only part of the HID API (set_idle, get_idle, sof) separately, even a prefix would be nice, just as an indication of transport dependent support for these.

It is USB-only part, it is prefixed with `usbd_`. However it evolves, there is no reason to add another prefix to anything. A callback is either optional or mandatory, it depends on the device. Adding prefixes does not change that.

benedekkupper replied:
> I removed */4U, it is 4 millisecond resolution now.

I actually preferred having ms resolution, but I'm also [vehemently against](https://github.com/UltimateHackingKeyboard/firmware/issues/390#issuecomment-991797166) the whole idle logic, so as you see fit.

> It is USB-only part, it is prefixed with usbd_.

The header is prefixed, but the API itself (types and methods) isn't, and also the explaining comment hinted at that this is supposed to be transport-independent. Moving on:

How about a suspend - resume callback, I'd see it as more relevant than having a sof callback? On the other hand these events are best handled on a whole device and not on an interface level.

jfischer-no replied:
> > I removed */4U, it is 4 millisecond resolution now.
> 
> I actually preferred having ms resolution, but I'm also [vehemently against](https://github.com/UltimateHackingKeyboard/firmware/issues/390#issuecomment-991797166) the whole idle logic, so as you see fit.

Changed it back and added comments about duration time resolution.
IMHO Get/Set Idle is an accident. Bluetooth was right to drop it.

> How about a suspend - resume callback, I'd see it as more relevant than having a sof callback? On the other hand these events are best handled on a whole device and not on an interface level.

There is common notification mechanism for the bus events. I also tested HID with suspend/rwup-requests (L2->L0), there will be a follow up PR to enable it for all functions. Also note that this API is experimental; if any limitations become apparent, we can fix them quickly.

MarekPieta commented on samples/subsys/usb/hid-keyboard/src/main.c:
```diff
@@ -0,0 +1,247 @@
+		LOG_ERR("Failed to enable device support");
+		return ret;
+	}
+
```
An info log informing that sample successfully initialized would be really helpful here. Currently there is only a boot banner if I boot without USB connected.

MarekPieta commented on samples/subsys/usb/hid-keyboard/src/main.c:
```diff
@@ -0,0 +1,247 @@
+			continue;
+		}
+
+		ret = hid_device_submit_report(hid_dev, sizeof(report), report, true);
```
By default, main thread is preemptive. USB HID device readiness may change after the `!kb_ready` check (it could also bypass internal checks done by `hid_device_submit_report` in similar way). Is it safe or could it lead to problems or edge cases in USB stack? Maybe we need cooperative priority here

jfischer-no replied:
Why do you think the cooperative thread can not be interrupted and is somehow different then?

MarekPieta replied:
A preemptive thread can be supplanted at any time by the scheduler while a cooperative thread needs to perform an action that would make it unready. As long as `input_cb` is called from a thread context (e.g. not from an interrupt), it would not preempt execution of a cooperative thread.

jfischer-no replied:
cooperative thread can be interrupted by an interrupt, AFAIK there is no guaranty it can immediately continue after, am I wrong?

tmon-nordic replied:
Cooperative thread will continue execution after it is interrupted by an interrupt. There is guarantee that there won't be other thread executing, see https://docs.zephyrproject.org/3.6.0/kernel/services/scheduling/index.html#cooperative-time-slicing for more information.

MarekPieta replied:
The `CONFIG_INPUT_MODE` is `CONFIG_INPUT_MODE_THREAD` by default, so that events seem broadcasted through a common input thread by default (then cooperative main priority could actually prevent supplant).

Ref: https://docs.zephyrproject.org/latest/services/input/index.html#application-api

jfischer-no replied:
I thought synchronous is on by default, added `CONFIG_INPUT_MODE_SYNCHRONOUS=y`.

jfischer-no replied:
> Cooperative thread will continue execution after it is interrupted by an interrupt. There is guarantee that there won't be other thread executing,

It seems to me that this article does not pay attention to meta-IRQ threads.

MarekPieta replied:
I wonder if using a thread for synchronizing inputs wouldn't actually simplify synchronization here (we could rely on cooperative priorities to simplify sample's implementation and prevent races). You are right that meta-IRQ threads are tricky use-case here (still, I don't recall seeing them used instead of regular threads in most of the subsystems or samples - I guess using them would require some "special handling" to ensure safety)

MarekPieta replied:
Still, since it's an improvement to the sample code, I would suggest to add it (but I would not NACK PR because of it)

MarekPieta commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+
+	return buf;
+}
+
```
Remove extra empty line, please

MarekPieta commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+	struct usbd_class_data *c_data = ddata->c_data;
+	struct net_buf *buf;
+
+
```
Remove extra empty line, please

MarekPieta commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+		return usbd_ep_buf_free(uds_ctx, buf);
+	}
+
+	if (sync && ops->input_report_done == NULL) {
```
Shouldn't it be `if (sync)` ? Currently it seems misaligned with API doc

jfischer-no replied:
I updated the documentation, but kept the sync parameter for now. I am still thinking about it.

MarekPieta replied:
Shouldn't it be `if (sync || ops->input_report_done == NULL)` then? (to align with newly introduced API doc)
Currently, if the callback is NULL and sync is set to false, the application has no way to get information that report was actually sent.

MarekPieta replied:
I see it's already aligned. Thanks

jfischer-no replied:
>Maybe we could simply use sync only if ops->input_report_done == NULL and use async otherwise (then an additional function argument may not be needed here).

I followed this suggestion.

tmon-nordic commented on dts/bindings/usb/zephyr,hid-device.yaml:
```diff
@@ -0,0 +1,56 @@
+      - keyboard: Device supports boot interface and keyboard protocol
+      - mouse:    Device supports boot interface and mouse protocol
+    enum:
+      - none
```
`none` should be removed from here. I know this requires extra macro mapping, but implicitly relying on numbers being the same between different types is something I prefer to avoid.

jfischer-no replied:
I did not get it. What types are different and how does removing "none" fix "implicitly relying on numbers being the same between different types" ?

tmon-nordic replied:
`none` doesn't really belong under "boot-protocol" from devicetree point of view IMHO (the absence of property signifies that). The "different types" I refer to the enum defined here in devicetree and the protocol code as defined by HID Specification - these two are essentially separate.

MarekPieta commented on subsys/usb/device_next/class/usbd_hid_api.c:
```diff
@@ -0,0 +1,220 @@
+
+int usb_hid_init(const struct device *dev)
+{
+	LOG_DBG("It does nothing for dev %p", dev);
```
Please cast the `%p` argument to `(void *)` (as required by C standard). Also in other places

jfischer-no replied:
Required by what C standard?
I will not add explicit conversion of any pointer to a void pointer.

tmon-nordic replied:
Apparently all of the major ones. C89: "The argument shall be a pointer to void", C99 and C11: "The corresponding argument shall be a pointer to a pointer to void.", C11

Not respecting this leads to undefined behavior. Whether this is going to lead to actual real world problems is another story, but from the standard point of view it can (as any undefined behavior).

carlescufi replied:
The other way around is OK. The C standard guarantees conversion from `void*` to any type of pointer, but not in this direction as pointed out here.

jfischer-no replied:
>The other way around is OK. The C standard guarantees conversion from void* to any type of pointer, but not in this direction as pointed out here.

That is different. C in a Nutshell, "Any object pointer type can be implicitly converted to a pointer to void, and vice versa.".
Not sure fprintf conversion specification is required for cbprintf.

MarekPieta commented on subsys/usb/device_next/class/usbd_hid_api.c:
```diff
@@ -0,0 +1,220 @@
+	int l = len;
+
+	if (legacy_ops != NULL && legacy_ops->set_report != NULL) {
+		return legacy_ops->get_report(dev, &setup, &l, &d);
```
Shouldn't it be `set_report` here?

MarekPieta commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+	if (ops->set_idle != NULL) {
+		ops->set_idle(dev, id, duration * 4UL);
+	} else {
+		errno = -ENOTSUP;
```
We shouldn't end up with this error if `id == 0U`

jfischer-no replied:
Why not? Id 0 applies to all reports and Set Report is not always required, so we need to issue protocol error if there is no callback.

MarekPieta replied:
Then shouldn't we also return error for `id == 0` in `handle_get_idle` if there is no callback? (for consistency).
Also, currently we update our common idle rate even though we return protocol error here.

btw. This is `set_idle`

jfischer-no replied:
Then I have to drop legacy API support.

MarekPieta replied:
I think it's worth to keep the legacy API support (even if it has some additional limitations)

MarekPieta commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+	switch (type) {
+	case HID_REPORT_TYPE_INPUT:
+		LOG_DBG("Set Report, Input Report ID %u", id);
+		errno = ops->set_report(dev, type, id, buf->len, buf->data);
```
Maybe we could make the `HID_REPORT_TYPE_` enum part of the application API and use it to pass report type?

jfischer-no commented on subsys/usb/device_next/class/usbd_hid_api.c:
```diff
@@ -0,0 +1,220 @@
+#include <zephyr/devicetree.h>
+#include <zephyr/usb/usbd.h>
+#include <zephyr/usb/class/usb_hid.h>
+#include <zephyr/usb/class/usbd_hid.h>
```
No

MarekPieta replied:
You are right, sorry for confusion (I misread the second one...)

jfischer-no commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+		duration = ops->get_idle(dev, id) / 4UL;
+	}
+
+	LOG_DBG("Get Idle, Report ID %u Duration %u", id, duration);
```
I do not find it practical, then we might also have to log every protocol error and that bloats everything.

MarekPieta replied:
I would not insist on adding extra logs, we can keep it as is

jfischer-no commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+	const uint8_t type = HID_GET_REPORT_TYPE(setup->wValue);
+	const uint8_t id = HID_GET_REPORT_ID(setup->wValue);
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
```
It makes no sense here, Get Report is mandatory for all devices, you cannot register a device without it.

MarekPieta replied:
You are right, a device may not use set report. We can keep it as is

jfischer-no commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+		ret = handle_set_protocol(dev, setup);
+		break;
+	default:
+		errno = -ENOTSUP;
```
Because it is protocol error.

tmon-nordic replied:
Returning error value is used to signify unrecoverable stack errors, while `errno` is used for protocol errors.

MarekPieta replied:
Thanks a lot for clarification

jfischer-no commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+	const struct hid_device_ops *const ops = ddata->ops;
+	struct usbd_hid_descriptor *const desc = ddata->desc;
+
+	atomic_set_bit(&ddata->state, HID_DEV_CLASS_ENABLED);
```
I do not understand what you are getting at and it is hard to guess. What are your current concerns?

MarekPieta replied:
E.g. `hid_dev_submit_report` may be called by an application from any context (also preemptive). Then it is possible that function execution would be preempted after `HID_DEV_CLASS_ENABLED` flag check and `usbd_ep_enqueue` would be called after HID device class was actually disabled. Is it acceptable?

MarekPieta replied:
Also the `hid_dev_output_handler` may be preempted in similar way (if someone uses preemptive system workqueue priority)

jfischer-no replied:
> Then it is possible that function execution would be preempted after `HID_DEV_CLASS_ENABLED` flag check and `usbd_ep_enqueue` would be called after HID device class was actually disabled. Is it acceptable?

Yes, it will fail and return an error.

MarekPieta replied:
Ok, so an application must handle this case if needed.

jfischer-no commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+	atomic_set_bit(&ddata->state, HID_DEV_CLASS_ENABLED);
+	ddata->protocol = HID_PROTOCOL_REPORT;
+	if (ops->iface_ready) {
+		ops->iface_ready(dev, true);
```
You should not mix them, HID device is one part, whatever else you have related to USB device management is another. iface_ready() is there for convenience, even with a message type like "configured:num" you would need to check, depending on the USB configuration, if it applies to the configuration and speed the HID device instance belongs to or not.

MarekPieta replied:
I understand that these are 2 separate things.

Still, maybe we could think about ensuring `iface_ready` notification order in relation to `UDC_EVT_`. It may be a bit confusing for an application if it e.g. gets `iface_ready` before `USBD_MSG_VBUS_READY` (e.g. because of blocked system workqueue). Or if resume from suspend is delayed for noticeable time after `iface_ready` was already called. Maybe we could e.g. keep the `iface_ready` API as is and just call application's callback through the system workqueue to ensure proper order?

jfischer-no replied:
No. Again, HID device iface_ready is low level, for the users top half of the driver. Do not mix them. Suspended bus does not mean that HID device is not ready.

tmon-nordic replied:
It is perfectly fine for host to suspend HID device that supports Remote Wakeup for aggressive power saving. It is up to device to initiate the Remote Wakeup. It makes sense for device to initiate wakeup only if it has something to send to host and therefore suspend/resume should not affect `iface_ready`.

MarekPieta replied:
Thanks for clarification

jfischer-no commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+	ddata->rsize = rsize;
+	ddata->ops = ops;
+
+	sys_put_le16(ddata->rsize, (uint8_t *)&(desc->hid.sub[0].wDescriptorLength));
```
rsize parameter is assign to ddata->rsize for convenience and then sub descriptor length is updated.

tmon-nordic replied:
It is not assigned twice, note that the second argument to `sys_put_le16()` is the destination and the first one is source value.

MarekPieta replied:
Sorry, I overlooked it. Looks good

jfischer-no commented on subsys/usb/device_next/class/usbd_hid_api.c:
```diff
@@ -0,0 +1,220 @@
+	ARG_UNUSED(dev);
+	ARG_UNUSED(proto_code);
+
+	LOG_DBG("Protocol code is set using DT property boot-protocol-code");
```
But it is not an error, it just there to get it compiled.

MarekPieta replied:
I mean that an application can no longer rely on this function after it switches to USB next stack (it actually does nothing here). It would be great if we could let user know about it. Debug log would not be displayed in most of the default app configurations I think (it may also be easily overlooked). Maybe we could e.g. change it to: `LOG_WRN_ONCE` ?

jfischer-no replied:
This it not called in a loop, so why should we use LOG_WRN_ONCE? I changed it to LOG_WRN().

MarekPieta replied:
You are right that there should not be too many USB HID class instances. `LOG_WRN` sounds good

jfischer-no commented on dts/bindings/usb/zephyr,hid-device.yaml:
```diff
@@ -16,7 +16,6 @@ properties:
 
   boot-protocol-code:
     type: string
-    required: true
```
Yes, I insist.

jfischer-no commented on samples/subsys/usb/hid-keyboard/src/main.c:
```diff
@@ -0,0 +1,247 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(main, LOG_LEVEL_INF);
+
+static const uint8_t __aligned(4) hid_report_desc[] = HID_KEYBOARD_REPORT_DESC();
```
I spotted it twice myself, but forgot to fix it.

MarekPieta replied:
Thanks for aligning it

jfischer-no commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+	usbd_ep_enqueue(c_data, buf);
+}
+
+static int hid_dev_submit_report(const struct device *dev,
```
No, one always have to enqueue a transfer.

MarekPieta replied:
Ok, thanks for clarification.

jfischer-no commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,743 @@
+		}
+	}
+
+	if (bi->ep == hid_get_in_ep(c_data)) {
```
Why would anyone do that?

MarekPieta replied:
I was thinking about explicitly avoiding the second comparison in code (`bi->ep` cannot match both conditions here)

jfischer-no replied:
I still do not get it.

tmon-nordic reviewed (dismissed):
I'd prefer `boot-protocol` instead of `protocol-code` in the devicetree property, but that can be argued to be a style choice.

MarekPieta commented on samples/subsys/usb/hid-keyboard/src/main.c:
```diff
@@ -0,0 +1,247 @@
+static int kb_get_report(const struct device *dev,
+			 const uint8_t type, const uint8_t id, const uint16_t len,
+			 uint8_t *const buf)
+{
```
Then maybe let's make it at least as `LOG_WRN`? I think it would be good to indicate that it should actually be implemented, but it's not yet here (you can even add a `TODO` comment)

MarekPieta replied:
I also wonder a bit how would HID host react if we won't handle `get_report` request here. May be good to double check to ensure HID device would still work after it.

MarekPieta replied:
Since it's a sample, we can keep it "handled" with a warning log for now

MarekPieta commented on samples/subsys/usb/hid-keyboard/src/main.c:
```diff
@@ -0,0 +1,247 @@
+			     const uint8_t *const buf)
+{
+	LOG_HEXDUMP_DBG(buf, len, "o.r.");
+	kb_set_report(dev, 2U, 0U, len, buf);
```
I meant that most of the parameters passed to `kb_set_report` are actually unused and we may desist from passing them as function arguments because of that

jfischer-no replied:
kb_set_report is the callback implementation for set_report, I changed it a bit, maybe it makes it clearer.

MarekPieta replied:
I see you prefer to pass the parameters anyway. We can keep it as is.

MarekPieta commented on subsys/usb/device_next/class/usbd_hid.c:
```diff
@@ -0,0 +1,740 @@
+	}
+
+	if (usbd_ep_enqueue(c_data, buf)) {
+		LOG_ERR("Failed to enqueue buffer");
```
Should we free buffer on error here?

jfischer-no replied:
:fire: I will look into what happened here.

MarekPieta replied:
It is already fixed. Thanks

MarekPieta reviewed (approved):
Overall looks good. Added a few more improvement propositions (and one small doc alignment)
